hudi-hadoop-common/src/main/java/org/apache/hudi/common/util/HFileUtils.java

```diff
@@ -128,4 +165,89 @@ public HoodieFileFormat getFormat() {
+                                           List<HoodieRecord> records,
+                                           Schema writerSchema,
+                                           Schema readerSchema,
+                                           String keyFieldName,
+                                           Map<String, String> paramsMap) throws IOException {
+    Compression.Algorithm compressionAlgorithm = getHFileCompressionAlgorithm(paramsMap);
+    HFileContext context = new HFileContextBuilder()
+        .withBlockSize(DEFAULT_BLOCK_SIZE_FOR_LOG_FILE)
+        .withCompression(compressionAlgorithm)
+        .withCellComparator(new HoodieHBaseKVComparator())
+        .build();
+
+    Configuration conf = storageConf.unwrapAs(Configuration.class);
+    CacheConfig cacheConfig = new CacheConfig(conf);
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    FSDataOutputStream ostream = new FSDataOutputStream(baos, null);
+
+    // Use simple incrementing counter as a key
+    boolean useIntegerKey = !getRecordKey(records.get(0), readerSchema, keyFieldName).isPresent();
+    // This is set here to avoid re-computing this in the loop
+    int keyWidth = useIntegerKey ? (int) Math.ceil(Math.log(records.size())) + 1 : -1;
+
+    // Serialize records into bytes
+    Map<String, List<byte[]>> sortedRecordsMap = new TreeMap<>();
+
+    Iterator<HoodieRecord> itr = records.iterator();
+    int id = 0;
+    while (itr.hasNext()) {
+      HoodieRecord<?> record = itr.next();
+      String recordKey;
+      if (useIntegerKey) {
+        recordKey = String.format("%" + keyWidth + "s", id++);
+      } else {
+        recordKey = getRecordKey(record, readerSchema, keyFieldName).get();
+      }
+
+      final byte[] recordBytes = serializeRecord(record, writerSchema, keyFieldName);
+      // If key exists in the map, append to its list. If not, create a new list.
+      // Get the existing list of recordBytes for the recordKey, or an empty list if it doesn't exist
+      List<byte[]> recordBytesList = sortedRecordsMap.getOrDefault(recordKey, new ArrayList<>());
+      recordBytesList.add(recordBytes);
+      // Put the updated list back into the map
+      sortedRecordsMap.put(recordKey, recordBytesList);
+    }
+
+    HFile.Writer writer = HFile.getWriterFactory(conf, cacheConfig)
+        .withOutputStream(ostream).withFileContext(context).create();
+
+    // Write the records
+    sortedRecordsMap.forEach((recordKey, recordBytesList) -> {
+      for (byte[] recordBytes : recordBytesList) {
+        try {
+          KeyValue kv = new KeyValue(recordKey.getBytes(), null, null, recordBytes);
+          writer.append(kv);
+        } catch (IOException e) {
+          throw new HoodieIOException("IOException serializing records", e);
+        }
+      }
+    });
+
+    writer.appendFileInfo(
+        getUTF8Bytes(HoodieAvroHFileReaderImplBase.SCHEMA_KEY), getUTF8Bytes(readerSchema.toString()));
+
+    writer.close();
+    ostream.flush();
```

Comment:
Wouldn't we want to flush before closing the writer?

Reply:
This is flushing the data to the `ByteArrayOutputStream` after the writer, and `write.close()` flushes the data internally.  This PR only moves this part of code from `HoodieHFileDataBlock` to `HFileUtils` class only.